code/game/atoms.dm

```diff
@@ -746,20 +746,41 @@
 		if(LAZYLEN(managed_vis_overlays))
 			SSvis_overlays.remove_vis_overlay(src, managed_vis_overlays)
 
-		var/list/new_overlays = update_overlays(updates)
-		if (managed_overlays)
-			if (length(overlays) == (islist(managed_overlays) ? length(managed_overlays) : 1))
-				overlays = null
-				POST_OVERLAY_CHANGE(src)
-			else
-				cut_overlay(managed_overlays)
-				managed_overlays = null
-		if(length(new_overlays))
-			if (length(new_overlays) == 1)
-				managed_overlays = new_overlays[1]
-			else
-				managed_overlays = new_overlays
-			add_overlay(new_overlays)
+		var/list/unknown_new_overlays = update_overlays(updates)
+		var/list/new_overlays = list()
+		for(var/atom/maybe_not_an_atom as anything in unknown_new_overlays)
+			if(isnull(maybe_not_an_atom))
+				continue
+			if(istext(maybe_not_an_atom) || isicon(maybe_not_an_atom))
+				new_overlays += maybe_not_an_atom
+				continue
+			new_overlays += maybe_not_an_atom.appearance
+
+		var/identical = FALSE
+		if(!islist(managed_overlays))
+			if(length(new_overlays) == 1 && managed_overlays == new_overlays[1])
+				identical = TRUE
+		else if(length(managed_overlays) == length(new_overlays))
+			identical = TRUE
+			for(var/i in 1 to length(managed_overlays))
+				if(managed_overlays[i] != new_overlays[i])
+					identical = FALSE
+					break
+
+		if(!identical)
+			if (managed_overlays)
+				if (length(overlays) == (islist(managed_overlays) ? length(managed_overlays) : 1))
+					overlays = null
+					POST_OVERLAY_CHANGE(src)
+				else
+					cut_overlay(managed_overlays)
+					managed_overlays = null
+			if(length(new_overlays))
+				if (length(new_overlays) == 1)
+					managed_overlays = new_overlays[1]
+				else
+					managed_overlays = new_overlays
+				add_overlay(new_overlays)
```

Comment:
part of me wants to say put build_appearance_list's shit in our loop here and make new procs that re add/cut overlay but without the extra work. we're already iterating.
That said, I hve recently come into the possesion of ideas for how to make this proc less of a problem. it's a no promises kinda thing but I have ideas dmn it

Reply:
This isn't in a loop here but, are you talking about unifying the appearance extraction code and making add_overlay assume all input is appearances?

Reply:
I mean the loop at 751, and build_appearance_list. add_overlay would still need to be as it is to support non appearance args, but a second proc would I think allow us to further optimize